shared/src/main/java/com/couchbase/lite/Database.java

```diff
@@ -1199,40 +1193,73 @@ private C4Document save(Document document, C4Document base, boolean deletion) th
         }
     }
 
-    /**
-     * "Pulls" from the database, merging the latest revision into the in-memory properties, without saving.
-     *
-     * @return Document[] 0 -> resolved, 1 -> current
-     */
-    private Document[] merge(Document document) throws CouchbaseLiteException {
-        // Read the current revision from the database:
-        Document current;
-        try {
-            current = new Document(this, document.getId(), true, true);
-        } catch (CouchbaseLiteException ex) {
-            if (ex.getCode() == Status.NotFound)
-                return null;
-            throw ex;
-        }
-
-        // Resolve conflict:
-        Document base = null;
-        if (document.getC4doc() != null)
-            base = new Document(this, document.getId(), document.getC4doc());
+    private boolean saveResolvedDocument(Document resolved, Conflict conflict) throws CouchbaseLiteException, LiteCoreException {
+        synchronized (lock) {
+            boolean commit = false;
+            beginTransaction();
+            try {
+                Document doc = conflict.getMine();
+                Document otherDoc = conflict.getTheirs();
 
-        ConflictResolver resolver = effectiveConflictResolver();
-        Conflict conflict = new Conflict(document, current, base);
+                // Figure out what revision to delete and what if anything to add:
+                String winningRevID;
+                String losingRevID;
+                byte[] mergedBody = null;
+                if (resolved == otherDoc) {
+                    winningRevID = otherDoc.getRevID();
+                    losingRevID = doc.getRevID();
+                } else {
+                    winningRevID = doc.getRevID();
+                    losingRevID = otherDoc.getRevID();
+                    if (resolved != doc) {
+                        resolved.setDatabase(this);
+                        mergedBody = resolved.encode();
+                        if (mergedBody == null)
+                            return false;
+                    }
+                }
 
-        Document resolved = resolver.resolve(conflict);
-        if (resolved == null)
-            throw new CouchbaseLiteException(LiteCoreDomain, kC4ErrorConflict);
+                // Tell LiteCore to do the resolution:
+                C4Document rawDoc = doc.getC4doc();
+                rawDoc.resolveConflict(winningRevID, losingRevID, mergedBody);
+                rawDoc.save(0);
+                Log.i(TAG, "Conflict resolved as doc '%s' rev %s", rawDoc.getDocID(), rawDoc.getRevID());
 
-        return new Document[]{resolved, current};
+                commit = true;
+            } finally {
+                endTransaction(commit);
+            }
+            return commit;
+        }
     }
 
     private ConflictResolver effectiveConflictResolver() {
         return getConflictResolver() != null ?
```

Comment:
conflict resolver shouldn't be null.

Reply:
`getConflictResolver()` returns from `DatabaseConfiguration.conflictResolver`. So it could be `null`. Or CBL Android implementation is incorrect.